src/neuroglancer/save_state/save_state.ts

```diff
@@ -0,0 +1,383 @@
+      button.classList.toggle('dirty', this.saves[this.key].dirty.value);
+    }
+  }
+
+  forceDirtyAsTrackable() {
+    if (!this.key) {
+      return;
+    }
+    const entry = this.saves[this.key];
+    if (entry.dirty === undefined || entry.dirty.value === undefined) {
+      this.saves[this.key].dirty = new TrackableBoolean(entry.dirty ? true : false);
+      this.saves[this.key].dirty.changed.add(this.unsaved.bind(this));
+      this.unsaved();
+    }
+  }
+
+  loadFromStorage() {
+    const params = new URLSearchParams(window.location.search);
+    this.key = <any> params.get('local_id');
+
+    if (this.key) {
+      location.hash = '';
+      const entry = this.saves[this.key];
+      if (entry) {
+        this.forceDirtyAsTrackable();
+        this.unsaved();
+        this.root.restoreState(entry.state);
+      } else {
+        StatusMessage.showTemporaryMessage(
+            `This URL is invalid. Do not copy the URL in the address bar. Use the save button.`,
+            10000);
+      }
+    }
+  }
+
+  updateStorage() {
+    let entry: SaveEntry;
+    if (!this.key) {
+      entry = recordEntry();
+      entry.dirty.changed.add(this.unsaved.bind(this));
+      this.key = entry.state_id;
+      this.saves[this.key] = entry;
+      this.unsaved();
+      const params = new URLSearchParams();
+      params.set('local_id', this.key);
+      history.pushState({}, '', `${window.location.origin}/?${params.toString()}`);
+    } else {
+      entry = this.saves[this.key];
+      this.forceDirtyAsTrackable();
+    }
+    const newState = this.root.toJSON();
+    if (JSON.stringify(entry.state) === JSON.stringify(newState)) {
+      return;
+    }
+    entry.state = this.root.toJSON();
+    entry.dirty.value = true;
+    this.push();
+  }
+
+  commit(source_url: string) {
+    if (this.key) {
+      this.savedUrl = source_url;
+      this.saves[this.key].dirty.value = false;
+      this.addToHistory(recordHistory(source_url));
+      this.push();
+    }
+  }
+
+  remoteLoad(json_url: string) {
+    StatusMessage.forPromise(
+        cancellableFetchOk(json_url, {}, responseJson).then(response => {
+          this.root.restoreState(response);
+        }),
+        {
+          initialMessage: `Retrieving state from json_url: ${json_url}.`,
+          delay: true,
+          errorPrefix: `Error retrieving state: `,
+        });
+  }
+
+  showSaveDialog(viewer: Viewer, jsonString?: string, get?: UrlType) {
+    new SaveDialog(viewer, jsonString, get);
+  }
+
+  showHistory(viewer: Viewer) {
+    new SaveHistoryDialog(viewer, this);
+  }
+}
+
+class SaveDialog extends Overlay {
+  constructor(public viewer: Viewer, jsonString?: string, get?: UrlType) {
+    super();
+
+    let urlStart = `${window.location.origin}${window.location.pathname}`;
+    let jsonUrl;
+    if (jsonString) {
+      jsonUrl = `${urlStart}?json_url=${jsonString}`;
+    } else {
+      if (!jsonUrl) {
+        jsonUrl = 'NOT AVALABLE';
```

Comment:
Typo: "available"
Maybe make it a constant to ensure that we are calling it the same everywhere in the code.